swim/memberlist.go

```diff
@@ -279,6 +281,83 @@ func (m *memberlist) SetLocalStatus(status string) {
 	m.postLocalUpdate()
 }
 
+func (m *memberlist) SetLocalLabel(key, value string) error {
+	// TODO implement a sane limit for the size of the labels to prevent users
+	// from impacting the performance of the gossip protocol.
+
+	// ensure that there is a labels map
+	if m.local.Labels == nil {
+		m.local.Labels = make(map[string]string)
+	}
+
+	// set the label
+	m.local.Labels[key] = value
+	m.postLocalUpdate()
+
+	// there was no error during the manipulation of labels
+	return nil
+}
+
+// GetLocalLabel returns the value of a label set on the local node. Its second
+// argument indicates if the key was present on the node or not
+func (m *memberlist) GetLocalLabel(key string) (string, bool) {
+	value, has := m.local.Labels[key]
+	return value, has
+}
+
+// LocalLabelsAsMap copies the labels set on the local node into a map for the
+// callee to use. Changes to this map will not be reflected in the labels kept
+// by this node.
+func (m *memberlist) LocalLabelsAsMap() map[string]string {
+	if len(m.local.Labels) == 0 {
+		return nil
+	}
+
+	cpy := make(map[string]string)
+	for k, v := range m.local.Labels {
+		cpy[k] = v
+	}
+	return cpy
+}
+
+// SetLocalLabels updates multiple labels at once. It will take all the labels
+// that are set in the map passed to this function and overwrite the value with
+// the value in the map. Keys that are not present in the provided map will
+// remain in the labels of this node.
+func (m *memberlist) SetLocalLabels(labels map[string]string) {
+	// ensure that there is a labels map
+	if m.local.Labels == nil {
+		m.local.Labels = make(map[string]string)
+	}
+
+	// copy the key-value pairs to our internal labels. By not setting the map
+	// of labels to the Labels value of the local member we prevent removing labels
+	// that the user did not specify in the new map.
+	for key, value := range labels {
+		m.local.Labels[key] = value
```

Comment:
In that case you should move the todo from `SetLocalLabel` to a higher level as well ;)